// nodes2ts/dist/S2Cell.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.S2Cell = void 0;
const Long = require("long");
const S2CellId_1 = require("./S2CellId");
const S2Point_1 = require("./S2Point");
const S2LatLng_1 = require("./S2LatLng");
const S2Projections_1 = require("./S2Projections");
const S2_1 = require("./S2");
const S2LatLngRect_1 = require("./S2LatLngRect");
const R1Interval_1 = require("./R1Interval");
const S1Interval_1 = require("./S1Interval");
const S2Cap_1 = require("./S2Cap");
class S2Cell {
    constructor(cellID) {
        this.cellID = cellID;
        if (cellID != null) {
            this.init(cellID);
        }
    }
    get id() {
        return this.cellID;
    }
    static fromFace(face) {
        return new S2Cell(S2CellId_1.S2CellId.fromFace(face));
    }
    // This is a static method in order to provide named parameters.
    static fromFacePosLevel(face, pos, level) {
        return new S2Cell(S2CellId_1.S2CellId.fromFacePosLevel(face, new Long(pos), level));
    }
    // Convenience methods.
    static fromPoint(p) {
        return new S2Cell(S2CellId_1.S2CellId.fromPoint(p));
    }
    static fromLatLng(ll) {
        return new S2Cell(S2CellId_1.S2CellId.fromPoint(ll.toPoint()));
    }
    isLeaf() {
        return this._level == S2CellId_1.S2CellId.MAX_LEVEL;
    }
    getVertex(k) {
        return S2Point_1.S2Point.normalize(this.getVertexRaw(k));
    }
    /**
     * Return the k-th vertex of the cell (k = 0,1,2,3). Vertices are returned in
     * CCW order. The points returned by GetVertexRaw are not necessarily unit
     * length.
     */
    getVertexRaw(k) {
        // Vertices are returned in the order SW, SE, NE, NW.
        return S2Projections_1.S2Projections.faceUvToXyz(this._face, ((k >> 1) ^ (k & 1)) == 0 ? this.uMin : this.uMax, (k >> 1) == 0 ? this.vMin : this.vMax);
    }
    getEdge(k) {
        return S2Point_1.S2Point.normalize(this.getEdgeRaw(k));
    }
    getEdgeRaw(k) {
        switch (k) {
            case 0:
                return S2Projections_1.S2Projections.getVNorm(this._face, this.vMin); // South
            case 1:
                return S2Projections_1.S2Projections.getUNorm(this._face, this.uMax); // East
            case 2:
                return S2Point_1.S2Point.neg(S2Projections_1.S2Projections.getVNorm(this._face, this.vMax)); // North
            default:
                return S2Point_1.S2Point.neg(S2Projections_1.S2Projections.getUNorm(this._face, this.uMin)); // West
        }
    }
    /**
     * Return the inward-facing normal of the great circle passing through the
     * edge from vertex k to vertex k+1 (mod 4). The normals returned by
     * GetEdgeRaw are not necessarily unit length.
     *
     *  If this is not a leaf cell, set children[0..3] to the four children of
     * this cell (in traversal order) and return true. Otherwise returns false.
     * This method is equivalent to the following:
     *
     *  for (pos=0, id=child_begin(); id != child_end(); id = id.next(), ++pos)
     * children[i] = S2Cell(id);
     *
     * except that it is more than two times faster.
     */
    subdivide() {
        // This function is equivalent to just iterating over the child cell ids
        // and calling the S2Cell constructor, but it is about 2.5 times faster.
        if (this.id.isLeaf()) {
            return null;
        }
        const children = new Array(4);
        for (let i = 0; i < 4; ++i) {
            children[i] = new S2Cell();
        }
        // Create four children with the appropriate bounds.
        let id = this.id.childBegin();
        const mid = this.getCenterUV();
        const uMid = mid.x;
        const vMid = mid.y;
        for (let pos = 0; pos < 4; ++pos, id = id.next()) {
            const child = children[pos];
            child._face = this.face;
            child._level = this.level + 1;
            child._orientation = this.orientation ^ S2_1.S2.POS_TO_ORIENTATION[pos];
            child.cellID = id;
            // We want to split the cell in half in "u" and "v".  To decide which
            // side to set equal to the midpoint value, we look at cell's (i,j)
            // position within its parent.  The index for "i" is in bit 1 of ij.
            const ij = S2_1.S2.POS_TO_IJ[this.orientation][pos];
            // The dimension 0 index (i/u) is in bit 1 of ij.
            if ((ij & 0x2) != 0) {
                child.uMin = uMid;
                child.uMax = this.uMax;
            }
            else {
                child.uMin = this.uMin;
                child.uMax = uMid;
            }
            // The dimension 1 index (j/v) is in bit 0 of ij.
            if ((ij & 0x1) != 0) {
                child.vMin = vMid;
                child.vMax = this.vMax;
            }
            else {
                child.vMin = this.vMin;
                child.vMax = vMid;
            }
        }
        return children;
    }
    /**
     * Return the direction vector corresponding to the center in (s,t)-space of
     * the given cell. This is the point at which the cell is divided into four
     * subcells; it is not necessarily the centroid of the cell in (u,v)-space or
     * (x,y,z)-space. The point returned by GetCenterRaw is not necessarily unit
     * length.
     */
    getCenter() {
        return S2Point_1.S2Point.normalize(this.getCenterRaw());
    }
    getCenterRaw() {
        return this.cellID.toPointRaw();
    }
    /**
     * Return the center of the cell in (u,v) coordinates (see {@code
     * S2Projections}). Note that the center of the cell is defined as the point
     * at which it is recursively subdivided into four children; in general, it is
     * not at the midpoint of the (u,v) rectangle covered by the cell
     */
    getCenterUV() {
        return this.cellID.getCenterUV();
    }
    /**
     * Return the average area of cells at this level. This is accurate to within
     * a factor of 1.7 (for S2_QUADRATIC_PROJECTION) and is extremely cheap to
     * compute.
     */
    static averageArea(level) {
        return S2Projections_1.S2Projections.AVG_AREA.getValue(level);
    }
    /**
     * Return the average area of cells at this level. This is accurate to within
     * a factor of 1.7 (for S2_QUADRATIC_PROJECTION) and is extremely cheap to
     * compute.
     */
    averageArea() {
        return S2Projections_1.S2Projections.AVG_AREA.getValue(this._level);
    }
    /**
     * Return the approximate area of this cell. This method is accurate to within
     * 3% percent for all cell sizes and accurate to within 0.1% for cells at
     * level 5 or higher (i.e. 300km square or smaller). It is moderately cheap to
     * compute.
     */
    approxArea() {
        // All cells at the first two levels have the same area.
        if (this._level < 2) {
            return this.averageArea();
        }
        // First, compute the approximate area of the cell when projected
        // perpendicular to its normal. The cross product of its diagonals gives
        // the normal, and the length of the normal is twice the projected area.
        const flatArea = S2Point_1.S2Point.crossProd(S2Point_1.S2Point.sub(this.getVertex(2), this.getVertex(0)), S2Point_1.S2Point.sub(this.getVertex(3), this.getVertex(1))).norm() * 0.5;
        // double flatArea = 0.5 * S2Point.crossProd(
        //         S2Point.sub(getVertex(2), getVertex(0)), S2Point.sub(getVertex(3), getVertex(1))).norm();
        // Now, compensate for the curvature of the cell surface by pretending
        // that the cell is shaped like a spherical cap. The ratio of the
        // area of a spherical cap to the area of its projected disc turns out
        // to be 2 / (1 + sqrt(1 - r*r)) where "r" is the radius of the disc.
        // For example, when r=0 the ratio is 1, and when r=1 the ratio is 2.
        // Here we set Pi*r*r == flat_area to find the equivalent disc.
        return flatArea * 2 / (Math.sqrt((Math.min(flatArea * S2_1.S2.M_1_PI, 1) * -1) + 1) + 1);
    }
    //
    // /**
    //  * Return the area of this cell as accurately as possible. This method is more
    //  * expensive but it is accurate to 6 digits of precision even for leaf cells
    //  * (whose area is approximately 1e-18).
    //  */
    exactArea() {
        const v0 = this.getVertex(0);
        const v1 = this.getVertex(1);
        const v2 = this.getVertex(2);
        const v3 = this.getVertex(3);
        return S2_1.S2.area(v0, v1, v2) + (S2_1.S2.area(v0, v2, v3));
    }
    // //////////////////////////////////////////////////////////////////////
    // S2Region interface (see {@code S2Region} for details):
    getCapBound() {
        // Use the cell center in (u,v)-space as the cap axis. This vector is
        // very close to GetCenter() and faster to compute. Neither one of these
        // vectors yields the bounding cap with minimal surface area, but they
        // are both pretty close.
        //
        // It's possible to show that the two vertices that are furthest from
        // the (u,v)-origin never determine the maximum cap size (this is a
        // possible future optimization).
        const uv = this.getCenterUV();
        const center = S2Point_1.S2Point.normalize(S2Projections_1.S2Projections.faceUvToXyz(this._face, uv.x, uv.y));
        let cap = S2Cap_1.S2Cap.fromAxisHeight(center, 0);
        for (let k = 0; k < 4; ++k) {
            cap = cap.addPoint(this.getVertex(k));
        }
        return cap;
    }
    // 35.26 degrees
    getPoint(i, j) {
        return S2Projections_1.S2Projections.faceUvToXyz(this._face, i == 0 ? this.uMin : this.uMax, j == 0 ? this.vMin : this.vMax);
    }
    getRectBound() {
        if (this._level > 0) {
            // Except for cells at level 0, the latitude and longitude extremes are
            // attained at the vertices. Furthermore, the latitude range is
            // determined by one pair of diagonally opposite vertices and the
            // longitude range is determined by the other pair.
            //
            // We first determine which corner (i,j) of the cell has the largest
            // absolute latitude. To maximize latitude, we want to find the point in
            // the cell that has the largest absolute z-coordinate and the smallest
            // absolute x- and y-coordinates. To do this we look at each coordinate
            // (u and v), and determine whether we want to minimize or maximize that
            // coordinate based on the axis direction and the cell's (u,v) quadrant.
            const u = this.uMin + this.uMax;
            const v = this.vMin + this.vMax;
            const i = S2Projections_1.S2Projections.getUAxis(this._face).z == 0 ? (u < 0 ? 1 : 0) : (u > 0 ? 1 : 0);
            const j = S2Projections_1.S2Projections.getVAxis(this._face).z == 0 ? (v < 0 ? 1 : 0) : (v > 0 ? 1 : 0);
            const lat = R1Interval_1.R1Interval.fromPointPair(S2LatLng_1.S2LatLng.latitude(this.getPoint(i, j)).radians, S2LatLng_1.S2LatLng.latitude(this.getPoint(1 - i, 1 - j)).radians);
            const lng = S1Interval_1.S1Interval.fromPointPair(S2LatLng_1.S2LatLng.longitude(this.getPoint(i, 1 - j)).radians, S2LatLng_1.S2LatLng.longitude(this.getPoint(1 - i, j)).radians);
            // DBL_EPSILON
            return new S2LatLngRect_1.S2LatLngRect(lat, lng)
                .expanded(S2LatLng_1.S2LatLng.fromRadians(S2_1.S2.DBL_EPSILON, S2_1.S2.DBL_EPSILON))
                .polarClosure();
        }
        // The face centers are the +X, +Y, +Z, -X, -Y, -Z axes in that order.
        // assert (S2Projections.getNorm(face).get(face % 3) == ((face < 3) ? 1 : -1));
        switch (this._face) {
            case 0:
                return new S2LatLngRect_1.S2LatLngRect(new R1Interval_1.R1Interval(-S2_1.S2.M_PI_4, S2_1.S2.M_PI_4), new S1Interval_1.S1Interval(-S2_1.S2.M_PI_4, S2_1.S2.M_PI_4));
            case 1:
                return new S2LatLngRect_1.S2LatLngRect(new R1Interval_1.R1Interval(-S2_1.S2.M_PI_4, S2_1.S2.M_PI_4), new S1Interval_1.S1Interval(S2_1.S2.M_PI_4, 3 * S2_1.S2.M_PI_4));
            case 2:
                return new S2LatLngRect_1.S2LatLngRect(new R1Interval_1.R1Interval(S2Cell.POLE_MIN_LAT, S2_1.S2.M_PI_2), new S1Interval_1.S1Interval(-S2_1.S2.M_PI, S2_1.S2.M_PI));
            case 3:
                return new S2LatLngRect_1.S2LatLngRect(new R1Interval_1.R1Interval(-S2_1.S2.M_PI_4, S2_1.S2.M_PI_4), new S1Interval_1.S1Interval(3 * S2_1.S2.M_PI_4, -3 * S2_1.S2.M_PI_4));
            case 4:
                return new S2LatLngRect_1.S2LatLngRect(new R1Interval_1.R1Interval(-S2_1.S2.M_PI_4, S2_1.S2.M_PI_4), new S1Interval_1.S1Interval(-3 * S2_1.S2.M_PI_4, -S2_1.S2.M_PI_4));
            default:
                return new S2LatLngRect_1.S2LatLngRect(new R1Interval_1.R1Interval(-S2_1.S2.M_PI_2, -S2Cell.POLE_MIN_LAT), new S1Interval_1.S1Interval(-S2_1.S2.M_PI, S2_1.S2.M_PI));
        }
    }
    mayIntersectC(cell) {
        return this.cellID.intersects(cell.cellID);
    }
    contains(p) {
        // We can't just call XYZtoFaceUV, because for points that lie on the
        // boundary between two faces (i.e. u or v is +1/-1) we need to return
        // true for both adjacent cells.
        const uvPoint = S2Projections_1.S2Projections.faceXyzToUv(this._face, p);
        if (uvPoint == null) {
            return false;
        }
        return (uvPoint.x >= this.uMin
            && uvPoint.x <= this.uMax
            && uvPoint.y >= this.vMin
            && uvPoint.y <= this.vMax);
    }
    // The point 'p' does not need to be normalized.
    containsC(cell) {
        return this.cellID.contains(cell.cellID);
    }
    init(id) {
        this.cellID = id;
        this._face = id.face;
        const ijo = id.toIJOrientation();
        this._orientation = S2CellId_1.S2CellId.getOrientation(ijo);
        this._level = id.level();
        const i = S2CellId_1.S2CellId.getI(ijo);
        const j = S2CellId_1.S2CellId.getJ(ijo);
        const cellSize = id.getSizeIJ();
        this.uMin = S2Projections_1.S2Projections.ijToUV(i, cellSize);
        this.uMax = S2Projections_1.S2Projections.ijToUV(i + cellSize, cellSize);
        this.vMin = S2Projections_1.S2Projections.ijToUV(j, cellSize);
        this.vMax = S2Projections_1.S2Projections.ijToUV(j + cellSize, cellSize);
        // for (let d = 0; d < 2; ++d) {
        //   // Compute the cell bounds in scaled (i,j) coordinates.
        //   const sijLo = (ij[d].val & -cellSize) * 2 - S2Cell.MAX_CELL_SIZE;
        //   const sijHi = sijLo + cellSize * 2;
        //   const s = 1/S2Cell.MAX_CELL_SIZE;
        //   this._uv[d][0] = R2Vector.singleStTOUV(s * (sijLo))
        //   //S2Projections.stToUV((1.0 / S2Cell.MAX_CELL_SIZE) * sijLo);
        //   this._uv[d][1] = R2Vector.singleStTOUV(s * (sijHi));
        //   //S2Projections.stToUV((1.0 / S2Cell.MAX_CELL_SIZE) * sijHi);
        // }
    }
    get face() {
        return this._face;
    }
    get orientation() {
        return this._orientation;
    }
    get level() {
        return this._level;
    }
    // Return the latitude or longitude of the cell vertex given by (i,j),
    // where "i" and "j" are either 0 or 1.
    toString() {
        return "[" + this._face + ", " + this._level + ", " + this.orientation + ", " + this.cellID + "]";
    }
    toGEOJSON() {
        const coords = [this.getVertex(0), this.getVertex(1), this.getVertex(2), this.getVertex(3), this.getVertex(0)]
            .map(v => S2LatLng_1.S2LatLng.fromPoint(v))
            .map(v => ([v.lngDegrees, v.latDegrees]));
        // const rectJSON = this.getRectBound().toGEOJSON();
        return {
            type: 'Feature',
            geometry: {
                type: 'Polygon',
                coordinates: [coords]
            },
            properties: {},
            title: `Cell: ${this.id.toToken()} lvl: ${this._level}`
        };
        // rectJSON.title = `Cell: ${this.id.toToken()}`;
        // return rectJSON;
    }
}
exports.S2Cell = S2Cell;
S2Cell.MAX_CELL_SIZE = 1 << S2CellId_1.S2CellId.MAX_LEVEL;
// We grow the bounds slightly to make sure that the bounding rectangle
// also contains the normalized versions of the vertices. Note that the
// maximum result magnitude is Pi, with a floating-point exponent of 1.
// Therefore adding or subtracting 2**-51 will always change the result.
//   private static MAX_ERROR = S2.toDecimal(1.0).dividedBy(S2.toDecimal(new Long(1).shiftLeft(51).toString()));
S2Cell.MAX_ERROR = 1 / new Long(1).shiftLeft(51).toNumber();
// The 4 cells around the equator extend to +/-45 degrees latitude at the
// midpoints of their top and bottom edges. The two cells covering the
// poles extend down to +/-35.26 degrees at their vertices.
// adding kMaxError (as opposed to the C version) because of asin and atan2
// roundoff errors
S2Cell.POLE_MIN_LAT = Math.asin(Math.sqrt(1 / 3)) - S2Cell.MAX_ERROR;
